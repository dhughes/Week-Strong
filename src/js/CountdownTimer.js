import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Vbox } from './Box';
import styled from 'styled-components';

const Wrapper = Vbox.extend`
  & > * {
    text-align: center;
  }
`;

const Timer = styled.div`
  font-size: 65vw;
  padding: 0;
  line-height: 1;
`;

const AlertTimer = Timer.extend`
  color: ${props => props.theme.warning}
`;

class CountdownTimer extends Component {
  static propTypes = {
    topLabel: PropTypes.string,
    seconds: PropTypes.number.isRequired,
    bottomLabel: PropTypes.string,
    onTimeout: PropTypes.func.isRequired
  };

  constructor(props) {
    super(props);
    this.state = {
      seconds: props.seconds,
      interval: null
    };
  }

  componentDidMount() {
    if (this.state.seconds === 0) return;

    let interval = setInterval(() => {
      this.setState(previousState => ({ seconds: previousState.seconds - 1 }));
      if (this.state.seconds === 0) {
        clearInterval(this.state.interval);
        this.props.onTimeout();
      }
    }, 1000);

    this.setState({ interval });
  }

  componentWillUnmount() {
    clearInterval(this.state.interval);
  }

  render() {
    return (
      <Wrapper>
        {this.props.topLabel !== undefined ? <h3>{this.props.topLabel}</h3> : null}
        {this.state.seconds > 3 ? <Timer>{this.state.seconds}</Timer> : <AlertTimer>{this.state.seconds}</AlertTimer>}
        {this.props.bottomLabel !== undefined ? <h3>{this.props.bottomLabel}</h3> : null}
      </Wrapper>
    );
  }
}

CountdownTimer.defaultProps = {
  onTimeout: () => {}
};

export default CountdownTimer;
