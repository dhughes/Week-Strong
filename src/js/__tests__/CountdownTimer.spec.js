import React from 'react';
import { shallow } from 'enzyme';
import CountdownTimer from '../CountdownTimer.js';

test('Countdown renders', () => {
  const component = shallow(<CountdownTimer seconds={5} />);
  expect(component).toMatchSnapshot();
});

test('Countdown renders top label', () => {
  const component = shallow(<CountdownTimer topLabel="Top Label" seconds={5} />);
  const label = <h3>Top Label</h3>;

  expect(component.contains(label)).toEqual(true);
});

test('Countdown renders bottom label', () => {
  const component = shallow(<CountdownTimer bottomLabel="Bottom Label" seconds={5} />);
  const label = <h3>Bottom Label</h3>;

  expect(component.contains(label)).toEqual(true);
});

test('Countdown renders both labels', () => {
  const component = shallow(<CountdownTimer topLabel="Top Label" bottomLabel="Bottom Label" seconds={5} />);
  const topLabel = <h3>Top Label</h3>;
  const bottomLabel = <h3>Bottom Label</h3>;

  expect(component.contains(topLabel)).toEqual(true);
  expect(component.contains(bottomLabel)).toEqual(true);
});
