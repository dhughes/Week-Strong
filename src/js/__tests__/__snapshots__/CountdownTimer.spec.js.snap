// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Countdown renders 1`] = `
<styled.div>
  <styled.div>
    5
  </styled.div>
</styled.div>
`;
